library/java/org/chromium/net/impl/JavaUrlRequest.java

```diff
@@ -119,778 +120,796 @@
+      execute(new CheckedRunnable() {
+        @Override
+        public void run() throws Exception {
+          mCallback.onRedirectReceived(JavaUrlRequest.this, info, newLocationUrl);
+        }
+      });
+    }
+
+    void onResponseStarted(UrlResponseInfo info) {
+      execute(new CheckedRunnable() {
+        @Override
+        public void run() throws Exception {
+          if (mState.compareAndSet(/* expected= */ State.STARTED,
+                                   /* updated= */ State.AWAITING_READ)) {
+            mCallback.onResponseStarted(JavaUrlRequest.this, mUrlResponseInfo);
+          }
+        }
+      });
+    }
+
+    void onReadCompleted(final UrlResponseInfo info, final ByteBuffer byteBuffer) {
+      execute(new CheckedRunnable() {
+        @Override
+        public void run() throws Exception {
+          if (mState.compareAndSet(/* expected= */ State.READING,
+                                   /* updated= */ State.AWAITING_READ)) {
+            mCallback.onReadCompleted(JavaUrlRequest.this, info, byteBuffer);
+          }
+        }
+      });
+    }
+
+    void onCanceled(final UrlResponseInfo info) {
+      closeResponseChannel();
+      mUserExecutor.execute(new Runnable() {
+        @Override
+        public void run() {
+          try {
+            mCallback.onCanceled(JavaUrlRequest.this, info);
+          } catch (Exception exception) {
+            Log.e(TAG, "Exception in onCanceled method", exception);
+          }
+        }
+      });
+    }
+
+    void onSucceeded(final UrlResponseInfo info) {
+      mUserExecutor.execute(new Runnable() {
+        @Override
+        public void run() {
+          try {
+            mCallback.onSucceeded(JavaUrlRequest.this, info);
+          } catch (Exception exception) {
+            Log.e(TAG, "Exception in onSucceeded method", exception);
+          }
+        }
+      });
+    }
+
+    void onFailed(final UrlResponseInfo urlResponseInfo, final CronetException e) {
+      closeResponseChannel();
+      Runnable runnable = new Runnable() {
+        @Override
+        public void run() {
+          try {
+            mCallback.onFailed(JavaUrlRequest.this, urlResponseInfo, e);
+          } catch (Exception exception) {
+            Log.e(TAG, "Exception in onFailed method", exception);
+          }
+        }
+      };
+      try {
+        mUserExecutor.execute(runnable);
+      } catch (InlineExecutionProhibitedException wasDirect) {
+        if (mFallbackExecutor != null) {
+          mFallbackExecutor.execute(runnable);
+        }
       }
     }
   }
-}
 
-private void closeResponseChannel() {
-  mExecutor.execute(new Runnable() {
+  private void closeResponseChannel() {
+    mExecutor.execute(new Runnable() {
+      @Override
+      public void run() {
+        if (mResponseChannel != null) {
+          try {
+            mResponseChannel.close();
+          } catch (IOException e) {
+            e.printStackTrace();
```

Comment:
Is this still needed?

Reply:
Well, this is like that in the original code:

https://chromium.googlesource.com/chromium/src/+/refs/heads/main/components/cronet/android/java/src/org/chromium/net/impl/JavaUrlRequest.java#900

Given that the Java implementation is a fifth wheel, just ignoring silently is good enough.

Done.